### PR TITLE
Feature/auto delegate processor

### PR DESCRIPTION
**Pull request by fsamuel-bs**
**Goals (and why)**: There are several Forwarding classes on our codebase, and more will be created with the startup ordering work. This PR adds a `@AutoDelegate` annotation, that generates a delegate interface from the annotated interface.

**Implementation Description (bullets)**: Add an annotation processor, `AutoDelegateProcessor`.

**Concerns (what feedback would you like?)**:
- which other corner cases should be tested?
- which other tests should be added?
- do I need to refactor something to make the code clearer?

**Where should we start reviewing?**: AutoDelegateProcessor is the main 

**Priority (whenever / two weeks / yesterday)**: today or tomorrow, since it's blocking startup ordering work.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2260)

### CODE REVIEW

**Comment by fsamuel-bs** [comment]
I'm currently working on making the tests more readable / adding more tests, but overall comments are welcomely added.

File: atlasdb-processors-tests/build.gradle [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+apply from: "../gradle/shared.gradle"
+
+apply plugin: 'org.inferred.processors'
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    processor project(":atlasdb-processors")
```
Comment (nziebart) [inline atlasdb-processors-tests/build.gradle]: feels a little odd to have a separate project for tests. why can't this be in `src/test/java` in the main project?

File: atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterface.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.processors;
+
+public interface ChildInterface extends RandomInterface {
+    @Override
+    default void method1() {}
```
Comment (nziebart) [inline atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterface.java]: as discussed in person, we should name each of these according to the characteristics that we're testing (e.g., `overriddenMethod()`, `voidMethodWithNoArgs`, etc.)

File: atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterface.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.processors;
+
+public interface ChildInterface extends RandomInterface {
```
Comment (nziebart) [inline atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterface.java]: nit: would suggest `TestInterface` over `RandomInterface`

File: atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterfaceImpl.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.processors;
+
+@AutoDelegate(interfaceToExtend = ChildInterface.class)
+public class ChildInterfaceImpl {
```
Comment (nziebart) [inline atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterfaceImpl.java]: nit: would suggest `ChildInterfaceWrapper` (it's not really an impl)

File: atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegate.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.CLASS)
+public @interface AutoDelegate {
+    /**
+     * The id of the interface to be extend.
```
Comment (nziebart) [inline atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegate.java]: nit: `type` rather than `id`

File: atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+import com.google.common.collect.MapMaker;
+import com.google.common.collect.Sets;
+import com.squareup.javapoet.JavaFile;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.TypeName;
+import com.squareup.javapoet.TypeSpec;
+
+@AutoService(Processor.class)
+public final class AutoDelegateProcessor extends AbstractProcessor {
+    private static final ConcurrentMap<ProcessingEnvironment, Processor> registeredProcessors =
```
Comment (nziebart) [inline atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java]: maybe add a comment about why we need the weak keys/values

File: atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+    @Override
+    public SourceVersion getSupportedSourceVersion() {
+        return SourceVersion.RELEASE_8;
+    }
+
+    @Override
+    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+        if (abortProcessing.get() == Boolean.TRUE) {
+            // Another instance of AutoDelegate is running in the current processing environment.
```
Comment (nziebart) [inline atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java]: should be `AutoDelegateProcessor`

File: atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterface.java (hunk 3 above)
Comment (fsamuel-bs) [inline atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterface.java]: Done.

File: atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterface.java (hunk 2 above)
Comment (fsamuel-bs) [inline atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterface.java]: Done.

File: atlasdb-processors-tests/build.gradle (hunk 1 above)
Comment (fsamuel-bs) [inline atlasdb-processors-tests/build.gradle]: For some reason, I couldn't run the processor in the same package as it's source code. Maybe it's possible, but I didn't figure a way to do it.

File: atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterfaceImpl.java (hunk 4 above)
Comment (fsamuel-bs) [inline atlasdb-processors-tests/src/main/java/com/palantir/processors/ChildInterfaceImpl.java]: Done. Renamed this to Wrapper and the TestInterface to impl.

File: atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegate.java (hunk 5 above)
Comment (fsamuel-bs) [inline atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegate.java]: Done.

File: atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java (hunk 7 above)
Comment (fsamuel-bs) [inline atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java]: Done.

File: atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java (hunk 6 above)
Comment (fsamuel-bs) [inline atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java]: Done.

File: atlasdb-processors-tests/build.gradle (hunk 1 above)
Comment (chrisalice) [inline atlasdb-processors-tests/build.gradle]: The gradle-processors plugin doesn't support that use-case; nor do IDEs.

File: atlasdb-processors-tests/src/main/java/com/palantir/processors/TestInterfaceImpl.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.processors;
+
+@AutoDelegate(interfaceToExtend = TestInterface.class)
+public class TestInterfaceImpl implements AutoDelegate_TestInterface {
```
Comment (chrisalice) [inline atlasdb-processors-tests/src/main/java/com/palantir/processors/TestInterfaceImpl.java]: The convention is to generate `AutoDelegate_TestInterfaceImpl`. Otherwise you cannot delegate multiple interfaces in the future, and your generated classes clash.

Comment (fsamuel-bs) [inline atlasdb-processors-tests/src/main/java/com/palantir/processors/TestInterfaceImpl.java]: Yep. I'm going to an approach that's going to generate abstract classes when extending a class and interfaces with default methods if the `typeToExtend` is an interface. There are not a lot of changes to the code, and it's the use case that Java's Proxy doesn't support (and we need it on our codebase), so I'm going to support it.

File: atlasdb-processors/src/main/java/com/palantir/processors/TypeToExtend.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    private Boolean classMethodFilter(Element element) {
+        return element.getKind() == ElementKind.METHOD
+                && element.getModifiers().contains(Modifier.PUBLIC)
+                && !element.getModifiers().contains(Modifier.FINAL)
+                && !element.getModifiers().contains(Modifier.NATIVE)
+                && !element.getModifiers().contains(Modifier.STATIC)
+                && !element.getSimpleName().contentEquals("equals")
+                && !element.getSimpleName().contentEquals("toString");
```
Comment (gsheasby) [inline atlasdb-processors/src/main/java/com/palantir/processors/TypeToExtend.java]: What about `hashCode`?

File: atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateClassTests.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Test
+    public void staticMethodsAreNotGenerated() {
+        Set<String> staticMethods = Arrays.stream(TEST_CLASS_METHODS)
+                .filter(method -> Modifier.isStatic(method.getModifiers()))
+                .map(TestingUtils::methodToString)
+                .collect(Collectors.toSet());
+        Set<String> generatedMethods = TestingUtils.extractMethods(AutoDelegate_TestClass.class);
+
+        assertThat(generatedMethods, not(hasItems(staticMethods.toArray(new String[0]))));
```
Comment (fsamuel-bs) [inline atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateClassTests.java]: Done.

File: atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+        while (superclass.getKind() != TypeKind.NONE) {
+            classIterator = ProcessorUtils.extractType(typeUtils, superclass);
+            superclasses.add(classIterator);
+            superclass = classIterator.getSuperclass();
+        }
+
+        return superclasses;
+    }
+
+    private List<TypeElement> fetchSuperinterfaces(TypeElement baseInterface) {
```
Comment (fsamuel-bs) [inline atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java]: Done.

File: atlasdb-processors/src/main/java/com/palantir/processors/ProcessorUtils.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        try {
+            // Throws a MirroredTypeException if the type is not compiled.
+            Class typeClass = annotation.typeToExtend();
+            type = elementUtils.getTypeElement(typeClass.getCanonicalName());
+        } catch (MirroredTypeException mte) {
+            DeclaredType typeMirror = (DeclaredType) mte.getTypeMirror();
+            type = (TypeElement) typeMirror.asElement();
+        }
+
+        return type;
```
Comment (fsamuel-bs) [inline atlasdb-processors/src/main/java/com/palantir/processors/ProcessorUtils.java]: Done.

File: atlasdb-processors/src/main/java/com/palantir/processors/TypeToExtend.java (hunk 9 above)
Comment (fsamuel-bs) [inline atlasdb-processors/src/main/java/com/palantir/processors/TypeToExtend.java]: `Object.java`'s `hashCode` is native, so it's filtered by the NATIVE check above. But agreed that hashCode could also be filtered with `equals` and `toString`. Added.

File: atlasdb-processors/src/main/java/com/palantir/processors/TypeToExtend.java [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+        this.typePackage = typePackage;
+
+        List<ExecutableElement> allMethods = extractMethods(typeToExtend);
+        for (TypeElement supertype : supertypes) {
+            allMethods.addAll(extractMethods(supertype));
+        }
+
+        Map<String, ExecutableElement> unifiedMethods = allMethods
+                .stream()
+                .collect(Collectors.toMap(ExecutableElement::toString, Function.identity(), (name1, name2) -> name1));
```
Comment (fsamuel-bs) [inline atlasdb-processors/src/main/java/com/palantir/processors/TypeToExtend.java]: The `(name1, name2) -> name1` function is the conflict resolution for this hash map. It says that, in case of methods with the same signature, just choose the 'first' one, since they're both the same. Methods with the same signature will occur when a subclass overrides a method.

This guarantees that we have a map of 'uniqueMethodName -> method', and we store all unique methods on the variable `methods`. I've renamed some methods around this area and added some comments. Let me know if it's clearer.

Comment (gsheasby) [inline atlasdb-processors/src/main/java/com/palantir/processors/TypeToExtend.java]: Yep, the comment is helpful here - thanks!
